tealer/analyses/utils/stack_emulator.py

```diff
@@ -0,0 +1,451 @@
+    getbit, setbit, getbyte, setbyte,
+    substring s e, substring3,
+    extract, extract3, extract_uint16, extract_uint32, extract_uint64,
+    replace2, replace3,
+    base64_decode, json_ref,
+
+
+Cryptographic/Hash:
+    sha256, keccak256, sha512_256, sha3_256,
+    ed25519verify, ed25519verify_bare,
+    ecdsa_verify, ecdsa_pk_decompress, ecdsa_pk_recover
+    vrf_verify,
+
+
+Logging Operations:
+    log
+
+
+Unclear/recent(v8) instructions:
+    proto, frame_dig, frame_bury,
+"""
+
+from typing import List, Union, Dict, Type, Tuple
+from functools import lru_cache
+
+
+from tealer.teal.basic_blocks import BasicBlock
+from tealer.teal.instructions.instructions import (
+    Instruction,
+)
+
+
+class UnknownStackValue:
+    """Represent a unknown stack value."""
+
+    def __str__(self) -> str:
+        return "UnknownStackValue()"
+
+    def __repr__(self) -> str:
+        return "UnknownStackValue()"
+
+
+class KnownStackValue:
+    """Represent values which are known during the local stack emulation.
+
+    A stack value is represented using the instruction, arguments and the position.
+
+    instruction: Every stack value is a result of executing a instruction. instruction
+        property contains the reference to that instruction's object.
+    args: A instruction consumes values from the stack. `args` represents the list of stack values
+        which are consumed by `self.instruction` to produce the result(s).
+    position: A instruction might produce multiple values. position represents the index in the list
+        of produced values.
+
+    e.g
+
+    int 10
+    int 11
+    mulw
+
+    values pushed by each instruction are:
+        "int 10" -> KnownStackValue(ins = Int(10), args = [])
+        "int 11" -> KnownStackValue(ins = Int(11), args = [])
+        "mulw" ->
+            KnownStackValue(
+                ins = Mulw(),
+                args = [Int(10), Int(11)],
+                position = 0,
+            ),
+            KnownStackValue(
+                ins = Mulw(),
+                args = [Int(10), Int(11)],
+                position = 1,
+            )
+
+        Value which is pushed first will have position 0.
+    """
+
+    def __init__(self, ins: Instruction, args: List, position: int = 0) -> None:
+        self._ins = ins
+        self._args: List[Union[KnownStackValue, UnknownStackValue]] = args
+        self._position = position
+
+    @property
+    def instruction(self) -> Instruction:
+        return self._ins
+
+    @property
+    def args(self) -> List[Union["KnownStackValue", UnknownStackValue]]:
+        """List of stack values consumed by self.instruction.
+
+        self.args[0] is the last popped(first pushed) value from the stack.
+        self.args[-1] is the first popped value from the stack:
+            self.args[-1] is on top of the stack right before the execution
+            of self.instruction.
+        """
+        return self._args
+
+    @property
+    def position(self) -> int:
```

Comment:
Add documentation; The notion of `position` is a bit uncommon and require extra doc